libhpcalcs/src/prime_cmd.c

```diff
@@ -360,47 +364,124 @@ HPEXPORT int HPCALL calc_prime_r_recv_screen(calc_handle * handle, calc_screensh
+        uint32_t size = namelen + file->size + 12; // Size of the data plus something
+        uint32_t other_size = namelen + file->size + 6; // Also size of the data plus something
+        prime_vtl_pkt * pkt;
+
+        // Some text editors add the UTF-16LE BOM at the beginning of the file, but the SDKV0.30 firmware version chokes on it.
+        // Therefore, skip the BOM.
+        if (   (file->type == PRIME_TYPE_PRGM || file->type == PRIME_TYPE_NOTE)
+            && (file->data[0] == 0xFF && file->data[1] == 0xFE)
+           ) {
+            offset = 2;
+            size -= 2;
+            other_size -= 2;
+        }
+
+        pkt = prime_vtl_pkt_new(size + header_size); // Add size of the header.
         if (pkt != NULL) {
             uint8_t * ptr;
             uint16_t crc16;
-            uint32_t offset = 0;
-
-            // Some text editors add the UTF-16LE BOM at the beginning of the file, but the SDKV0.30 firmware version chokes on it.
-            // Therefore, skip the BOM.
-            if (   (file->type == PRIME_TYPE_PRGM || file->type == PRIME_TYPE_NOTE)
-                && (file->data[0] == 0xFF && file->data[1] == 0xFE)
-               ) {
-                offset = 2;
-                size -= 2;
-            }
 
             pkt->cmd = CMD_PRIME_RECV_FILE;
             ptr = pkt->data;
-            *ptr++ = CMD_PRIME_RECV_FILE;
+
+            // Command sequence. Connectivity kit increments this after each command, but the Prime seems to ignore it.
             *ptr++ = 0x01;
-            *ptr++ = (uint8_t)((size >> 24) & 0xFF);
-            *ptr++ = (uint8_t)((size >> 16) & 0xFF);
-            *ptr++ = (uint8_t)((size >>  8) & 0xFF);
+            *ptr++ = 0x00;
+            *ptr++ = 0x00;
+            *ptr++ = 0x00;  
+
             *ptr++ = (uint8_t)((size      ) & 0xFF);
+            *ptr++ = (uint8_t)((size >>  8) & 0xFF);
+            *ptr++ = (uint8_t)((size >> 16) & 0xFF);
+            *ptr++ = (uint8_t)((size >> 24) & 0xFF);
+
+            *ptr++ = CMD_PRIME_RECV_FILE;
+            
+            // ?
+            *ptr++ = 0x03;
+            
+            // Why not use different endiannesses for sizes within the same package... It's more fun that way.
+            *ptr++ = (uint8_t)((other_size >> 24) & 0xFF);
+            *ptr++ = (uint8_t)((other_size >> 16) & 0xFF);
+            *ptr++ = (uint8_t)((other_size >>  8) & 0xFF);
+            *ptr++ = (uint8_t)((other_size      ) & 0xFF);
+            
             *ptr++ = file->type;
+            
             *ptr++ = namelen;
-            *ptr++ = 0x00; // CRC16, set it to 0 for now.
+
+            // CRC16, set it to 0 for now.
+            *ptr++ = 0x00;
             *ptr++ = 0x00;
+
             memcpy(ptr, file->name, namelen);
             ptr += namelen;
+
             memcpy(ptr, file->data + offset, file->size - offset);
-            crc16 = crc16_block(pkt->data, size); // Yup, the last 6 bytes of the packet are excluded from the CRC.
```

Comment:
The new CRC includes all of the data (minus header), and no longer omits the last 6 bytes :)